include/grpcpp/impl/codegen/call.h

```diff
@@ -598,12 +792,353 @@ class CallOpSetInterface : public CompletionQueueTag {
+  // will be invoked. This is to be used only by BaseAsyncRequest and
+  // SyncRequest.
+  bool RunInterceptors(std::function<void(void)> f) {
+    // This is used only by the server for initial call request
+    GPR_CODEGEN_ASSERT(reverse_ == true);
+    GPR_CODEGEN_ASSERT(call_->client_rpc_info() == nullptr);
+    auto* server_rpc_info = call_->server_rpc_info();
+    if (server_rpc_info == nullptr ||
+        server_rpc_info->interceptors_.size() == 0) {
+      return true;
+    }
+    callback_ = std::move(f);
+    RunServerInterceptors();
+    return false;
+  }
+
+ private:
+  void RunClientInterceptors() {
+    auto* rpc_info = call_->client_rpc_info();
+    if (!reverse_) {
+      curr_iteration_ = 0;
+    } else {
+      if (rpc_info->hijacked_) {
+        curr_iteration_ = rpc_info->hijacked_interceptor_;
+      } else {
+        curr_iteration_ = rpc_info->interceptors_.size() - 1;
+      }
+    }
+    rpc_info->RunInterceptor(this, curr_iteration_);
+  }
+
+  void RunServerInterceptors() {
+    auto* rpc_info = call_->server_rpc_info();
+    if (!reverse_) {
+      curr_iteration_ = 0;
+    } else {
+      curr_iteration_ = rpc_info->interceptors_.size() - 1;
+    }
+    rpc_info->RunInterceptor(this, curr_iteration_);
+  }
+
+  void ProceedClient() {
+    auto* rpc_info = call_->client_rpc_info();
+    if (rpc_info->hijacked_ && !reverse_ &&
+        curr_iteration_ == rpc_info->hijacked_interceptor_ &&
+        !ran_hijacking_interceptor_) {
+      // We now need to provide hijacked recv ops to this interceptor
+      ClearHookPoints();
+      ops_->SetHijackingState();
+      ran_hijacking_interceptor_ = true;
+      rpc_info->RunInterceptor(this, curr_iteration_);
+      return;
+    }
+    if (!reverse_) {
+      curr_iteration_++;
+      // We are going down the stack of interceptors
+      if (curr_iteration_ < static_cast<long>(rpc_info->interceptors_.size())) {
+        if (rpc_info->hijacked_ &&
+            curr_iteration_ > rpc_info->hijacked_interceptor_) {
+          // This is a hijacked RPC and we are done with hijacking
+          ops_->ContinueFillOpsAfterInterception();
+        } else {
+          rpc_info->RunInterceptor(this, curr_iteration_);
+        }
+      } else {
+        // we are done running all the interceptors without any hijacking
+        ops_->ContinueFillOpsAfterInterception();
+      }
+    } else {
+      curr_iteration_--;
+      // We are going up the stack of interceptors
+      if (curr_iteration_ >= 0) {
+        // Continue running interceptors
+        rpc_info->RunInterceptor(this, curr_iteration_);
+      } else {
+        // we are done running all the interceptors without any hijacking
+        ops_->ContinueFinalizeResultAfterInterception();
+      }
+    }
+  }
+
+  void ProceedServer() {
+    auto* rpc_info = call_->server_rpc_info();
+    if (!reverse_) {
+      curr_iteration_++;
+      if (curr_iteration_ < static_cast<long>(rpc_info->interceptors_.size())) {
```

Comment:
I suggest that all array index calculations, comparisons, etc should be done as `size_t`

Reply:
done!

Reply:
current_interceptor_index I have kept as a long to allow it going negative. I could change that too if you think that's better.